File: source/georef/street_network.cpp
/* Copyright © 2001-2022, Hove and/or its affiliates. All rights reserved.

This file is part of Navitia,
    the software to build cool stuff with public transport.

Hope you'll enjoy and contribute to this project,
    powered by Hove (www.hove.com).
Help us simplify mobility and open public transport:
    a non ending quest to the responsive locomotion way of traveling!

LICENCE: This program is free software; you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <http://www.gnu.org/licenses/>.

Stay tuned using
twitter @navitia
channel `#navitia` on riot https://riot.im/app/#/room/#navitia:matrix.org
https://groups.google.com/d/forum/navitia
www.navitia.io
*/

#include "street_network.h"

#include "georef.h"
#include "type/data.h"

#include <chrono>

namespace navitia {
namespace georef {

StreetNetwork::StreetNetwork(const GeoRef& geo_ref)
    : geo_ref(geo_ref), departure_path_finder(geo_ref), arrival_path_finder(geo_ref), direct_path_finder(geo_ref) {}

void StreetNetwork::init(const type::EntryPoint& start, const boost::optional<const type::EntryPoint&>& end) {
    departure_path_finder.init(start.coordinates, start.streetnetwork_params.mode,
                               start.streetnetwork_params.speed_factor);
    if (end) {
        arrival_path_finder.init((*end).coordinates, (*end).streetnetwork_params.mode,
                                 (*end).streetnetwork_params.speed_factor);
    }
}

bool StreetNetwork::departure_launched() const {
    return departure_path_finder.computation_launch;
}
bool StreetNetwork::arrival_launched() const {
    return arrival_path_finder.computation_launch;
}

routing::map_stop_point_duration StreetNetwork::find_nearest_stop_points(
    const navitia::time_duration& radius,
    const proximitylist::ProximityList<type::idx_t>& pl,
    bool use_second) {
    // delegate to the arrival or departure pathfinder
    // results are store to build the routing path after the transportation routing computation
    return (use_second ? arrival_path_finder : departure_path_finder).find_nearest_stop_points(radius, pl);
}

navitia::time_duration StreetNetwork::get_distance(type::idx_t target_idx, bool use_second) {
    return (use_second ? arrival_path_finder : departure_path_finder).get_distance(target_idx);
}

Path StreetNetwork::get_path(type::idx_t idx, bool use_second) {
    Path result;
    if (!use_second) {
        result = departure_path_finder.get_path(idx);

        if (!result.path_items.empty()) {
            result.path_items.front().coordinates.push_front(departure_path_finder.starting_edge.projected);
        }
    } else {
        result = arrival_path_finder.get_path(idx);

        // we have to reverse the path
        std::reverse(result.path_items.begin(), result.path_items.end());
        int last_angle = 0;
        const auto& speed_factor = arrival_path_finder.speed_factor;
        for (auto& item : result.path_items) {
            std::reverse(item.coordinates.begin(), item.coordinates.end());

            // we have to reverse the directions too
            // the first direction become 0,
            // and we 'shift' all directions to the next path_item after reverting them
            int current_angle = -1 * item.angle;
            item.angle = last_angle;
            last_angle = current_angle;

            // FIXME: ugly temporary fix
            // while we don't use a boost::reverse_graph, the easiest way to handle
            // the bss rent/putback section in the arrival section is to swap them
            if (item.transportation == PathItem::TransportCaracteristic::BssTake) {
                item.transportation = PathItem::TransportCaracteristic::BssPutBack;
                item.duration = geo_ref.default_time_bss_putback / speed_factor;
            } else if (item.transportation == PathItem::TransportCaracteristic::BssPutBack) {
                item.transportation = PathItem::TransportCaracteristic::BssTake;
                item.duration = geo_ref.default_time_bss_pickup / speed_factor;
            }
            if (item.transportation == PathItem::TransportCaracteristic::CarPark) {
                item.transportation = PathItem::TransportCaracteristic::CarLeaveParking;
                item.duration = geo_ref.default_time_parking_leave / speed_factor;
            } else if (item.transportation == PathItem::TransportCaracteristic::CarLeaveParking) {
                item.transportation = PathItem::TransportCaracteristic::CarPark;
                item.duration = geo_ref.default_time_parking_park / speed_factor;
            }
        }

        if (!result.path_items.empty()) {
            // no direction for the first elt
            result.path_items.back().coordinates.push_back(arrival_path_finder.starting_edge.projected);
        }
    }

    return result;
}

Path StreetNetwork::get_direct_path(const type::EntryPoint& origin, const type::EntryPoint& destination) {
    auto dest_mode = origin.streetnetwork_params.mode;
    if (dest_mode == type::Mode_e::Car) {
        // on direct path with car we want to arrive on the walking graph
        dest_mode = type::Mode_e::Walking;
    }
    const auto dest_edge = ProjectionData(destination.coordinates, geo_ref, dest_mode);
    if (!dest_edge.found) {
        return Path();
    }
    const auto max_dur = origin.streetnetwork_params.max_duration + destination.streetnetwork_params.max_duration;
    direct_path_finder.init(origin.coordinates, dest_edge.projected, origin.streetnetwork_params.mode,
                            origin.streetnetwork_params.speed_factor);

    direct_path_finder.start_distance_or_target_astar(max_dur, dest_edge.projected,
                                                      {dest_edge[source_e], dest_edge[target_e]});
    const auto dest_vertex = direct_path_finder.find_nearest_vertex(dest_edge, true);
    const auto res = direct_path_finder.get_path(dest_edge, dest_vertex);
    if (res.duration > max_dur) {
        return Path();
    }
    return res;
}
}  // namespace georef
}  // namespace navitia

File: source/jormungandr/jormungandr/street_network/street_network.py
# coding=utf-8

# Copyright (c) 2001-2016, Hove and/or its affiliates. All rights reserved.
#
# This file is part of Navitia,
#     the software to build cool stuff with public transport.
#
# Hope you'll enjoy and contribute to this project,
#     powered by Hove (www.hove.com).
# Help us simplify mobility and open public transport:
#     a non ending quest to the responsive locomotion way of traveling!
#
# LICENCE: This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
# Stay tuned using
# twitter @navitia
# channel `#navitia` on riot https://riot.im/app/#/room/#navitia:matrix.org
# https://groups.google.com/d/forum/navitia
# www.navitia.io
from __future__ import absolute_import, print_function, unicode_literals, division

from jormungandr import utils, new_relic

import abc
from enum import Enum
import six

# Using abc.ABCMeta in a way it is compatible both with Python 2.7 and Python 3.x
# http://stackoverflow.com/a/38668373/1614576
ABC = abc.ABCMeta(str("ABC"), (object,), {})


# Regarding to the type of direct path, some special treatments may be done in connector
class StreetNetworkPathType(Enum):
    DIRECT = 0
    BEGINNING_FALLBACK = 1
    ENDING_FALLBACK = 2


from collections import namedtuple

StreetNetworkPathKey = namedtuple(
    'StreetNetworkPathKey', ['mode', 'orig_uri', 'dest_uri', 'streetnetwork_path_type', 'period_extremity']
)


class AbstractStreetNetworkService(ABC):  # type: ignore
    def get_street_network_routing_matrix(
        self, instance, origins, destinations, street_network_mode, max_duration, request, request_id, **kwargs
    ):
        with new_relic.record_streetnetwork_call(
            "routing_matrix", type(self).__name__, street_network_mode, instance.name
        ):
            return self._get_street_network_routing_matrix(
                instance, origins, destinations, street_network_mode, max_duration, request, request_id, **kwargs
            )

    @abc.abstractmethod
    def _get_street_network_routing_matrix(
        self, instance, origins, destinations, street_network_mode, max_duration, request, request_id, **kwargs
    ):
        pass

    @abc.abstractmethod
    def status(self):
        pass

    def filter_places_isochrone(self, places_isochrone):
        """
        This method is used by geovelo to sort places_isochrone by mode weight
        :param places_isochrone:
        :return: sorted and truncated places_isochrone
        """
        return places_isochrone

    def direct_path_with_fp(
        self,
        instance,
        mode,
        pt_object_origin,
        pt_object_destination,
        fallback_extremity,
        request,
        direct_path_type,
        request_id,
    ):
        with new_relic.record_streetnetwork_call("direct_path", type(self).__name__, mode, instance.name):
            resp = self._direct_path(
                instance,
                mode,
                pt_object_origin,
                pt_object_destination,
                fallback_extremity,
                request,
                direct_path_type,
                request_id,
            )

        return resp

    @abc.abstractmethod
    def _direct_path(
        self,
        instance,
        mode,
        pt_object_origin,
        pt_object_destination,
        fallback_extremity,
        request,
        direct_path_type,
        request_id,
    ):
        '''
        :param fallback_extremity: is a PeriodExtremity (a datetime and it's meaning on the fallback period)
        :param direct_path_type : direct_path need to be treated differently regarding to the used connector
        '''
        pass

    @abc.abstractmethod
    def make_path_key(self, mode, orig_uri, dest_uri, streetnetwork_path_type, period_extremity):
        """
        This method is used for the caching method. It's connector specific.
        :param orig_uri, dest_uri, mode: matters obviously
        :param streetnetwork_path_type: whether it's a fallback at the beginning,
            the end of journey or a direct path without PT also matters especially for car (to know if we park before or after)
        :param period_extremity: is a PeriodExtremity (a datetime and it's meaning on the fallback period)

        """
        pass

    def feed_publisher(self):
        return None

    def record_external_failure(self, message):
        utils.record_external_failure(message, 'streetnetwork', six.text_type(self.sn_system_id))

    def record_call(self, status, **kwargs):
        """
        status can be in: ok, failure
        """
        params = {'streetnetwork_id': six.text_type(self.sn_system_id), 'status': status}
        params.update(kwargs)
        new_relic.record_custom_event('streetnetwork', params)

    def _add_feed_publisher(self, resp):
        sn_feed = self.feed_publisher()
        if sn_feed:
            feed = resp.feed_publishers.add()
            feed.id = sn_feed.id
            feed.name = sn_feed.name
            feed.license = sn_feed.license
            feed.url = sn_feed.url
